src/box/memtx_tree.cc

```diff
@@ -1013,77 +1013,86 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
  * Implementation of iterator position for general and multikey indexes.
  */
 template <bool USE_HINT, bool IS_MULTIKEY>
-static int
-tree_iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+static inline int
+tree_iterator_position_impl(struct memtx_tree_data<USE_HINT> *last,
+			    struct index_def *def,
+			    const char **pos, uint32_t *size)
 {
 	static_assert(!IS_MULTIKEY || USE_HINT,
 		      "Multikey index actually uses hint.");
-	struct memtx_tree_index<USE_HINT> *index =
-		(struct memtx_tree_index<USE_HINT> *)it->index;
-	struct key_def *cmp_def = index->base.def->cmp_def;
-	struct tree_iterator<USE_HINT> *tree_it =
-		get_tree_iterator<USE_HINT>(it);
-	struct tuple *tuple = tree_it->last.tuple;
+	struct tuple *tuple = last->tuple;
 	if (tuple == NULL) {
 		*pos = NULL;
 		*size = 0;
 		return 0;
 	}
 	int mk_idx = MULTIKEY_NONE;
 	if (IS_MULTIKEY)
-		mk_idx = (int)tree_it->last.hint;
-	const char *key = tuple_extract_key(tuple, cmp_def, mk_idx, size);
+		mk_idx = last->hint;
+	const char *key = tuple_extract_key(tuple, def->cmp_def, mk_idx, size);
 	if (key == NULL)
 		return -1;
 	*pos = key;
 	return 0;
 }
 
+/**
+ * Implementation of iterator position for general and multikey indexes.
+ */
+template <bool USE_HINT, bool IS_MULTIKEY>
+static int
+tree_iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+{
+	static_assert(!IS_MULTIKEY || USE_HINT,
+		      "Multikey index actually uses hint.");
+	struct memtx_tree_index<USE_HINT> *index =
+		(struct memtx_tree_index<USE_HINT> *)it->index;
+	struct tree_iterator<USE_HINT> *tree_it =
+		get_tree_iterator<USE_HINT>(it);
+	return tree_iterator_position_impl<USE_HINT, IS_MULTIKEY>(
+		&tree_it->last, index->base.def, pos, size);
+}
+
 /**
  * Implementation of iterator position for functional indexes.
  */
 static int
-tree_iterator_position_func(struct iterator *it, const char **pos,
-			    uint32_t *size)
+tree_iterator_position_func_impl(struct memtx_tree_data<true> *last,
+				 struct index_def *def,
+				 const char **pos, uint32_t *size)
 {
 	/*
 	 * Сmp_def in the functional index is the functional key and the
 	 * primary key right after it. So to extract cmp_def in func index,
 	 * we need to pack an array with concatenated func key and primary key.
 	 */
-	struct tree_iterator<true> *tree_it = get_tree_iterator<true>(it);
-	if (tree_it->last_func_key == NULL) {
+	if (last->tuple == NULL) {
 		*pos = NULL;
 		*size = 0;
 		return 0;
 	}
-	assert(tree_it->last.tuple != NULL);
 	/* Extract func key. */
 	uint32_t func_key_size;
-	const char *func_key = tuple_data_range(tree_it->last_func_key,
+	struct tuple *func_key_tup = (struct tuple *)last->hint;
+	const char *func_key = tuple_data_range(func_key_tup,
 						&func_key_size);
 	uint32_t func_key_len = mp_decode_array(&func_key);
 	/* Extract primary key. */
-	struct space *space = space_by_id(it->index->def->space_id);
+	struct space *space = space_by_id(def->space_id);
```

Comment:
You can't use `space_by_id` for read views because a space captured by a read view may be dropped.

Reply:
Already in master